cluster.go

```diff
@@ -1056,16 +1131,94 @@ func (c *Cluster) pin(pin api.Pin, blacklist []peer.ID, prioritylist []peer.ID)
 // of underlying IPFS daemon unpinning operations.
 func (c *Cluster) Unpin(h *cid.Cid) error {
 	logger.Info("IPFS cluster unpinning:", h)
+	cState, err := c.consensus.State()
+	if err != nil {
+		return err
+	}
 
-	pin := api.Pin{
-		Cid: h,
+	if !cState.Has(h) {
+		return errors.New("cannot unpin pin uncommitted to state")
+	}
+	pin := cState.Get(h)
+
+	switch pin.Type {
+	case api.DataType:
+		return c.consensus.LogUnpin(pin)
+	case api.ShardType:
+		err := "unpinning shard cid %s before unpinning parent"
```

Comment:
Glad you asked this.  It is possible right now in the case of a cluster failure during sharding.  Because sharding is currently approaching a prototype that will not have any data safety guarantees or builtin recovery mechanisms this might be useful in the near future.  The alternative will be cleaning up a failed sharding session by cleaning up the state and starting fresh.  I am happy to change this to a warning if you prefer.  Might also be nice to include the `-f` flag into the semantics of unpin so that certain pin types can only be unpinned if done with the force flag.

Reply:
oh god did I ask you to disallow it on the first review and now I was asking you to re-allow it?

Reply:
haha not exactly. Last review you said "unsure we should allow this at all" as a comment to the warning I had in place.